# Census Population Data

The goal of this notebook is to pull in census data and make it available to incorporate into the dashboard. Specifically to do cases per capita and deaths per capita. 

Change directory to parent directory, so we can make use of modules

In [1]:
cd ..

/Users/DanOvadia/Projects/covid-hotspots


### Import Modules

In [2]:
import pandas as pd

In [3]:
from modules import data_processing
from modules import plotting

Data come from census.gov from their 2019 population estimates for all counties, available for direct download. The csv is encoded using `encoding = "ISO-8859-1"`.

 - [Webpage](https://www2.census.gov/programs-surveys/popest/datasets/2010-2019/counties/totals/)
 - [Direct Download CSV (3.5MB)](https://www2.census.gov/programs-surveys/popest/datasets/2010-2019/counties/totals/co-est2019-alldata.csv)

In [4]:
URL = 'https://www2.census.gov/programs-surveys/popest/datasets/2010-2019/counties/totals/co-est2019-alldata.csv'
CENSUS_DF = pd.read_csv(URL, encoding = "ISO-8859-1")

In [5]:
CENSUS_DF.shape

(3193, 164)

In [6]:
CENSUS_DF.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3193 entries, 0 to 3192
Data columns (total 164 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   SUMLEV                 int64  
 1   REGION                 int64  
 2   DIVISION               int64  
 3   STATE                  int64  
 4   COUNTY                 int64  
 5   STNAME                 object 
 6   CTYNAME                object 
 7   CENSUS2010POP          int64  
 8   ESTIMATESBASE2010      int64  
 9   POPESTIMATE2010        int64  
 10  POPESTIMATE2011        int64  
 11  POPESTIMATE2012        int64  
 12  POPESTIMATE2013        int64  
 13  POPESTIMATE2014        int64  
 14  POPESTIMATE2015        int64  
 15  POPESTIMATE2016        int64  
 16  POPESTIMATE2017        int64  
 17  POPESTIMATE2018        int64  
 18  POPESTIMATE2019        int64  
 19  NPOPCHG_2010           int64  
 20  NPOPCHG_2011           int64  
 21  NPOPCHG_2012           int64  
 22  NPOPCHG_2013           

In [7]:
CENSUS_DF.head(10)

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME          CTYNAME  \
0      40       3         6      1       0  Alabama          Alabama   
1      50       3         6      1       1  Alabama   Autauga County   
2      50       3         6      1       3  Alabama   Baldwin County   
3      50       3         6      1       5  Alabama   Barbour County   
4      50       3         6      1       7  Alabama      Bibb County   
5      50       3         6      1       9  Alabama    Blount County   
6      50       3         6      1      11  Alabama   Bullock County   
7      50       3         6      1      13  Alabama    Butler County   
8      50       3         6      1      15  Alabama   Calhoun County   
9      50       3         6      1      17  Alabama  Chambers County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2019  \
0        4779736            4780125          4785437  ...          1.917501   
1          54571              54597            54773  ...          4.847310   
2         182265             182265           183112  ...         24.017829   
3          27457              27455            27327  ...         -5.690302   
4          22915              22915            22870  ...          1.385134   
5          57322              57322            57376  ...          1.020788   
6          10914              10911            10876  ...         -7.102343   
7          20947              20940            20932  ...         -7.216152   
8         118572             118526           118408  ...         -4.167837   
9          34215              34169            34122  ...         -7.927723   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  \
0     0.578434     1.186314     1.522549     0.563489     0.626357   
1     6.018182    -6.226119    -3.902226     1.970443    -1.712875   
2    16.641870    17.488579    22.751474    20.184334    17.725964   
3     0.292676    -6.897817    -8.132185    -5.140431   -15.724575   
4    -4.998356    -3.787545    -5.797999     1.331144     1.329817   
5     0.208812    -1.650165    -0.347225    -2.049590    -1.338525   
6   -21.994339    -6.954560    -6.334200     9.617198   -24.592888   
7    -3.636538    -7.896764   -14.478623     0.442445    -6.223913   
8    -6.148582    -4.588523    -5.255477    -4.027747    -3.215406   
9    -1.379209     4.490952     2.520405    -3.787656     1.324055   

   RNETMIG2016  RNETMIG2017  RNETMIG2018  RNETMIG2019  
0     0.745172     1.090366     1.773786     2.483744  
1     4.777171     0.849656     0.540916     4.560062  
2    21.279291    22.398256    24.727215    24.380567  
3   -18.238016   -24.998528    -8.754922    -5.165664  
4    -0.708717    -3.234669    -6.857092     1.831952  
5    -1.391062     6.193562    -0.069229     1.124597  
6    -2.212709   -19.936786    -1.474201    -7.200986  
7    -6.272714    -4.962406   -10.071105    -6.294941  
8    -3.471589    -1.654454    -0.488995    -4.044995  
9    -5.786747     2.134851    -1.515444    -7.748227  

[10 rows x 164 columns]

Selecing a handful of columns to persist; might alter as I better understand what this dataset has as a whole.

In [8]:
FEATURES = ['SUMLEV','REGION','DIVISION','STATE','COUNTY','STNAME','CTYNAME','POPESTIMATE2019','CENSUS2010POP']

In [9]:
CENSUS_DF[FEATURES].head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   POPESTIMATE2019  CENSUS2010POP  
0          4903185        4779736  
1            55869          54571  
2           223234         182265  
3            24686          27457  
4            22394          22915

--------

## FIPS Codes

In order to incorporate these data into the covid dashboard, we'll need to tie them to counties via FIPS codes. We have state and county codes available above as integers, but we'll need it stored as a 5 digit string.

Below I create a function to apply to the dataframe to produce the new column. This is open to issues if state or county codes exceed expected lengths.

In [10]:
# Function to generate fips from a state int and county int.
def generate_fips(state_fips, county_fips):
    state_str, county_str = str(state_fips), str(county_fips)
    
    # Check length of state code and append 0's if necessary.
    if len(state_str) == 1:
        state_str = "0"+state_str
        
    # Check length of county code and append 0's if necessary.
    if len(county_str) == 1:
        county_str = "00"+county_str
    elif len(county_str) == 2:
        county_str = "0"+county_str
        
    return state_str+county_str

# Quick test
state = 1
county = 22
generate_fips(state, county)

'01022'

In [11]:
# Use apply across rows to generate fips codes.
CENSUS_DF['FIPS'] = CENSUS_DF[['STATE','COUNTY']].apply(
    lambda x:generate_fips(
        x['STATE'],x['COUNTY']), axis = 1)

# Add the new column to our features.
if 'FIPS' not in FEATURES:
    FEATURES.append('FIPS')

### Length Check

Lets check to make sure that our transformation was clean from a length persepective.

In [12]:
# Checking the the length of each row, producing a series
#  and checking the max.
CENSUS_DF['FIPS'].map(lambda x:len(str(x))).max()

5

In [13]:
# Doing the same only for min.
CENSUS_DF['FIPS'].map(lambda x:len(str(x))).min()

5

So we've got 3193 rows (counties), lets make sure we have the same number of unique fips codes.

### Uniqueness Check

In [14]:
CENSUS_DF.shape

(3193, 165)

In [15]:
len(set(CENSUS_DF['FIPS']))

3193

Yup, we look good to go.

------------

In [16]:
COVID_COUNTY_DF = data_processing.get_covid_county_data()

Retrieving Covid County data
Pulling county data from github.


In [21]:
date_mask = (COVID_COUNTY_DF['date'] == '2020-08-28')

fips_error_mask = (COVID_COUNTY_DF['fips'].map(lambda x:len(x)) < 5)

COVID_COUNTY_DF[fips_error_mask & date_mask]

date         county          state fips   cases  deaths  fipsnum  \
476351 2020-08-28        Unknown         Alaska  nan      56       0      NaN   
476440 2020-08-28        Unknown       Arkansas  nan     815       0      NaN   
476574 2020-08-28        Unknown    Connecticut  nan     104       0      NaN   
476579 2020-08-28        Unknown       Delaware  nan     279       0      NaN   
476644 2020-08-28        Unknown        Florida  nan     959       0      NaN   
476793 2020-08-28        Unknown        Georgia  nan    2497       4      NaN   
476809 2020-08-28        Unknown           Guam  nan    2256      11      NaN   
476949 2020-08-28        Unknown       Illinois  nan    1645     207      NaN   
477141 2020-08-28        Unknown           Iowa  nan      16       0      NaN   
477434 2020-08-28        Unknown      Louisiana  nan     350     163      NaN   
477456 2020-08-28        Unknown          Maine  nan       3       0      NaN   
477481 2020-08-28        Unknown       Maryland  nan       0      17      NaN   
477498 2020-08-28        Unknown  Massachusetts  nan    1533       5      NaN   
477579 2020-08-28        Unknown       Michigan  nan     345       4      NaN   
477662 2020-08-28        Unknown      Minnesota  nan     196      49      NaN   
477805 2020-08-28         Joplin       Missouri  nan     691      21      NaN   
477806 2020-08-28    Kansas City       Missouri  nan    8995      96      NaN   
478009 2020-08-28        Unknown       Nebraska  nan      77       0      NaN   
478042 2020-08-28        Unknown  New Hampshire  nan       3       0      NaN   
478063 2020-08-28        Unknown     New Jersey  nan     328       0      NaN   
478125 2020-08-28  New York City       New York  nan  238107   23680      NaN   
478304 2020-08-28        Unknown   North Dakota  nan       4       4      NaN   
478471 2020-08-28        Unknown       Oklahoma  nan      21       0      NaN   
478650 2020-08-28        Unknown    Puerto Rico  nan    1328       0      NaN   
478662 2020-08-28        Unknown   Rhode Island  nan    1874      19      NaN   
478863 2020-08-28        Unknown      Tennessee  nan     206       0      NaN   
479146 2020-08-28        Unknown           Utah  nan      17       1      NaN   
479163 2020-08-28        Unknown        Vermont  nan       8       0      NaN   
479337 2020-08-28        Unknown     Washington  nan       2       2      NaN   
479460 2020-08-28        Unknown      Wisconsin  nan    4305       8      NaN   

        log_deaths  
476351    0.000000  
476440    0.000000  
476574    0.000000  
476579    0.000000  
476644    0.000000  
476793    1.609438  
476809    2.484907  
476949    5.337538  
477141    0.000000  
477434    5.099866  
477456    0.000000  
477481    2.890372  
477498    1.791759  
477579    1.609438  
477662    3.912023  
477805    3.091042  
477806    4.574711  
478009    0.000000  
478042    0.000000  
478063    0.000000  
478125   10.072428  
478304    1.609438  
478471    0.000000  
478650    0.000000  
478662    2.995732  
478863    0.000000  
479146    0.693147  
479163    0.000000  
479337    1.098612  
479460    2.197225